addons/test_mail/tests/test_ir_attachment.py

```diff
@@ -0,0 +1,77 @@
+from odoo.addons.mail.tests.common import mail_new_test_user
+from odoo.addons.test_mail.tests.common import TestMailCommon
+from odoo.exceptions import AccessError
+
+class TestAttachment(TestMailCommon):
+
+    def test_attachment_forbid_unlink(self):
+        """Check that removing message attachments is prevented on other user's messages."""
+        test_record = self.env['mail.test.simple'].with_context(self._test_context).create({
+            'name': 'Test',
+            'email_from': 'ignasse@example.com',
+        })
+
+        user_second_employee = mail_new_test_user(
+            self.env,
+            login="employee_second",
+            email="employee_second@example.com",
+            groups="base.group_user",
+        )
+
+        admin_attachments = self.env['ir.attachment'].with_user(self.user_admin).create([{
+            'type': 'binary',
+            'name': f'admin_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(2)])
+        employee_attachments = self.env['ir.attachment'].with_user(self.user_employee).create([{
+            'type': 'binary',
+            'name': f'employee_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(2)])
+
+        second_employee_attachment = self.env['ir.attachment'].with_user(user_second_employee).create({
+            'type': 'binary',
+            'name': 'second_employee_attachment',
+            'datas': "ABC=",
+        })
+
+        # used in different messages by different users
+        shared_attachment_employee = self.env['ir.attachment'].with_user(self.user_employee).create({
+            'res_model': test_record._name,
+            'res_id': test_record.id,
+            'type': 'binary',
+            'name': 'shared_attachment_employee',
+            'datas': "ABC=",
+        })
+
+        test_record.with_user(self.user_admin).message_post(body="Hi", attachment_ids=admin_attachments.ids)
+        test_record.with_user(self.user_employee).message_post(body="Hello", attachment_ids=(employee_attachments + shared_attachment_employee).ids)
+        test_record.with_user(user_second_employee).sudo().message_post(body="Hello again", attachment_ids=shared_attachment_employee.ids)
```

Comment:
Marked as resolved but I still see the sudo :eyes:

Reply:
Ah, but it's not required :smirk: 
Fixed.

Also unpublished the enterprise branch to make runbot happy, it was just for tracking.